src/emu/machine.cpp

```diff
@@ -1116,7 +1116,19 @@ std::string running_machine::nvram_filename(device_t &device) const
 		if (software != nullptr && *software != '\0')
 			result << PATH_SEPARATOR << software;
 
-		std::string tag(device.tag());
+		std::string tag;
+		for (device_t *dev = &device; dev->owner() != nullptr; dev = dev->owner())
+		{
+			std::ostringstream dev_tag;
+			dev_tag << ":" << dev->basetag();
+			device_slot_interface *intf;
+			if (dev->owner() && dev->owner()->interface(intf))
+			{
+				if (dev->system_bios() != 0 && dev->default_bios() != dev->system_bios())
+					util::stream_format(dev_tag, "_%d", dev->system_bios() - 1);
+			}
+			tag = dev_tag.str() + tag;
+		}
```

Comment:
I don’t think using BIOS numbers in anything externally visible is a good idea – they a lot less stable than the mnemonic names.

Also, does the check actually work properly?  What happens when the default BIOS isn’t the first one (e.g. `osborne1`)?  It looks like both the default BIOS and the first BIOS will have no suffix.

Reply:
Please note that this part of code is affecting only bios on slot device level, and for those on driver, logic is not changed, that is in lines 1099-1100. 
For slot devices I have changed code according to comment.